File: frontend/ios/ElBotonWidget/BackendAPIClient.swift
import Foundation

class BackendAPIClient {
    static let shared = BackendAPIClient()

    private init() {}

    func createAlert(
        credentials: WidgetCredentials,
        latitude: Double,
        longitude: Double
    ) async throws -> String {
        let url = URL(string: "\(credentials.apiUrl)/alerts")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue("Bearer \(credentials.accessToken)", forHTTPHeaderField: "Authorization")

        let body: [String: Any] = [
            "latitude": latitude,
            "longitude": longitude
        ]

        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await URLSession.shared.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw AlertIntentError.apiError("Invalid response")
        }

        guard (200...299).contains(httpResponse.statusCode) else {
            let errorBody = String(data: data, encoding: .utf8) ?? "Unknown error"
            throw AlertIntentError.apiError("Status \(httpResponse.statusCode): \(errorBody)")
        }

        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let alertId = json["id"] as? String else {
            throw AlertIntentError.apiError("Failed to parse alert response")
        }

        return alertId
    }

    func refreshToken(credentials: WidgetCredentials) async throws -> (accessToken: String, refreshToken: String, expiresAt: TimeInterval) {
        let url = URL(string: "\(credentials.apiUrl)/auth/refresh")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("application/json", forHTTPHeaderField: "Accept")

        let body: [String: Any] = [
            "refresh_token": credentials.refreshToken
        ]
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await URLSession.shared.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse,
              (200...299).contains(httpResponse.statusCode) else {
            throw AlertIntentError.apiError("Token refresh failed")
        }

        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let accessToken = json["access_token"] as? String,
              let refreshToken = json["refresh_token"] as? String,
              let expiresIn = json["expires_in"] as? Int else {
            throw AlertIntentError.apiError("Failed to parse refresh response")
        }

        let expiresAt = Date().timeIntervalSince1970 + Double(expiresIn)
        return (accessToken, refreshToken, expiresAt)
    }
}

File: frontend/plugins/withAlertWidget/android/BackendApiClient.kt
package com.elboton.app.widget

import com.google.gson.Gson
import com.google.gson.JsonObject
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.util.concurrent.TimeUnit

class BackendApiClient {

    private val httpClient = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .writeTimeout(30, TimeUnit.SECONDS)
        .build()

    private val gson = Gson()

    fun createAlert(credentials: Credentials, latitude: Double, longitude: Double): String? {
        val body = mapOf(
            "latitude" to latitude,
            "longitude" to longitude
        )

        val jsonBody = gson.toJson(body)
        val mediaType = "application/json".toMediaType()
        val requestBody = jsonBody.toRequestBody(mediaType)

        val request = Request.Builder()
            .url("${credentials.apiUrl}/alerts")
            .post(requestBody)
            .addHeader("Content-Type", "application/json")
            .addHeader("Accept", "application/json")
            .addHeader("Authorization", "Bearer ${credentials.accessToken}")
            .build()

        return try {
            val response = httpClient.newCall(request).execute()

            if (response.isSuccessful) {
                val responseBody = response.body?.string()
                if (responseBody != null) {
                    val alertObj = gson.fromJson(responseBody, JsonObject::class.java)
                    alertObj.get("id")?.asString
                } else {
                    null
                }
            } else {
                null
            }
        } catch (e: Exception) {
            e.printStackTrace()
            null
        }
    }

    fun refreshToken(credentials: Credentials): RefreshTokenResult? {
        val body = mapOf(
            "refresh_token" to credentials.refreshToken
        )

        val jsonBody = gson.toJson(body)
        val mediaType = "application/json".toMediaType()
        val requestBody = jsonBody.toRequestBody(mediaType)

        val request = Request.Builder()
            .url("${credentials.apiUrl}/auth/refresh")
            .post(requestBody)
            .addHeader("Content-Type", "application/json")
            .addHeader("Accept", "application/json")
            .build()

        return try {
            val response = httpClient.newCall(request).execute()

            if (response.isSuccessful) {
                val responseBody = response.body?.string()
                if (responseBody != null) {
                    val result = gson.fromJson(responseBody, RefreshTokenResponse::class.java)
                    RefreshTokenResult(
                        accessToken = result.access_token,
                        refreshToken = result.refresh_token,
                        expiresAt = System.currentTimeMillis() / 1000 + result.expires_in
                    )
                } else {
                    null
                }
            } else {
                null
            }
        } catch (e: Exception) {
            e.printStackTrace()
            null
        }
    }

    data class RefreshTokenResponse(
        val access_token: String,
        val refresh_token: String,
        val expires_in: Long
    )

    data class RefreshTokenResult(
        val accessToken: String,
        val refreshToken: String,
        val expiresAt: Long
    )
}
